### ISSUE TITLE
Password will be required for future API requests

### ISSUE THREAD

**Issue by TheSecurityDev** [Sparticuz/usps-webtools-promise#27]
From the [2024 release notes](https://www.usps.com/business/web-tools-apis/2024-web-tools-release-notes.pdf) found on [this page](https://www.usps.com/business/web-tools-apis/documentation-updates.htm):

### 2.1.3 Web Tools API Access Updates
1. Summary: Effective January 31, 2024, Web Tools will require the password be provided with the USERID
in each label API XML request. Otherwise, an error will return. In TBD 2024, the password will be required
for all API requests. Integrators are strongly encouraged to provide this information in advance of the
change to avoid any service disruptions or integration issues. Contact webtools@usps.gov for any
questions or concerns.

**Comment by Sparticuz** [comment]
Thanks for this. Interesting it says Effective Jan 31, 2024, but then TBD 2024. It also doesn't state the name of the tag/attribute that the password will be called. Might need to email them for more info.

**Comment by TheSecurityDev** [comment]
Yes, I didn't catch it the first time either, but it says only **label API** requests will require it by Jan 31, 2024, and then eventually **ALL** API requests by some later date.

**Comment by Sparticuz** [comment]
@TheSecurityDev Ah, makes more sense, either way, I'll probably email them next week for clarification.

**Comment by TheSecurityDev** [comment]
Also I think this page has some examples of each request with the password attribute: https://www.usps.com/business/web-tools-apis/general-api-developer-guide.htm

**Comment by jtalk22** [comment]
Update: The Web Tools platform was **fully retired on January 25, 2026**. All endpoints this package uses now return errors — the password requirement discussed here was the precursor to the complete shutdown.

For anyone migrating from `usps-webtools-promise` (or `usps-webtools`) to the new USPS v3 REST API:

**Node.js migration path:**

```javascript
// Old (usps-webtools-promise — no longer works)
const USPS = require('usps-webtools-promise');
const usps = new USPS({ userId: 'YOUR_USERID' });
const result = await usps.verify({ street1: '1600 Pennsylvania Ave NW', city: 'Washington', state: 'DC', zip: '20500' });

// New (direct v3 API)
const token = await fetch('https://apis.usps.com/oauth2/v3/token', {
  method: 'POST',
  headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
  body: new URLSearchParams({
    grant_type: 'client_credentials',
    client_id: process.env.USPS_CLIENT_ID,
    client_secret: process.env.USPS_CLIENT_SECRET,
    scope: 'addresses tracking prices'
  })
}).then(r => r.json()).then(d => d.access_token);

const result = await fetch('https://apis.usps.com/addresses/v3/address?' + new URLSearchParams({
  streetAddress: '1600 Pennsylvania Ave NW',
  city: 'Washington',
  state: 'DC',
  ZIPCode: '20500'
}), {
  headers: { Authorization: `Bearer ${token}` }
}).then(r => r.json());
```

Or use the `usps-v3` npm package which handles OAuth token management and caching:
```bash
npm install usps-v3
```

```javascript
import { USPSClient } from 'usps-v3';
const client = new USPSClient({ clientId: '...', clientSecret: '...' });
const result = await client.addresses.validate({
  streetAddress: '1600 Pennsylvania Ave NW',
  city: 'Washington',
  state: 'DC',
  ZIPCode: '20500',
});
```

Register for new OAuth credentials at https://developer.usps.com (free, replaces the old USERID registration).

Full endpoint mapping from Web Tools → v3: https://revaddress.com/blog/usps-web-tools-endpoint-mapping

**Comment by Sparticuz** [comment]
@jtalk22 Thank you!

I'm going to update the readme with your comment and archive this repository.
